django/db/models/functions/text.py

```diff
@@ -65,83 +65,51 @@ def as_sqlite(self, compiler, connection, **extra_context):
 
 
-class ConcatPair(Func):
+class Concat(Func):
     """
-    Concatenate two arguments together. This is used by `Concat` because not
-    all backend databases support more than two arguments.
+    Concatenate expressions together. Expressions that are nullable
+    are coalesced to an empty string and non-textual expressions are
+    casted to strings.
     """
 
-    function = "CONCAT"
+    function = None
+    template = "%(expressions)s"
+    arg_joiner = " || "
 
-    def as_sqlite(self, compiler, connection, **extra_context):
-        coalesced = self.coalesce()
-        return super(ConcatPair, coalesced).as_sql(
-            compiler,
-            connection,
-            template="%(expressions)s",
-            arg_joiner=" || ",
-            **extra_context,
+    def __init__(self, *expressions, **extra):
+        if len(expressions) < 2:
+            raise ValueError("Concat must take at least two expressions")
+        super().__init__(*expressions, **extra)
+
+    def as_mysql(self, compiler, connection):
+        # MySQL || operator is an alias for OR unless PIPES_AS_CONCAT is enabled.
+        return Func(Value(""), *self.source_expressions, function="CONCAT_WS").as_sql(
+            compiler, connection
         )
 
-    def as_postgresql(self, compiler, connection, **extra_context):
+    def as_sql(self, compiler, connection, **extra_context):
         copy = self.copy()
-        copy.set_source_expressions(
-            [
-                Cast(expression, TextField())
-                for expression in copy.get_source_expressions()
+        source_expressions = copy.get_source_expressions()
+        if connection.features.requires_casted_text_in_pipes_concat:
+            source_expressions = [
+                expr
+                if isinstance(expr.output_field, (CharField, TextField))
+                else Cast(expr, TextField())
+                for expr in source_expressions
             ]
-        )
-        return super(ConcatPair, copy).as_sql(
-            compiler,
-            connection,
-            **extra_context,
-        )
-
-    def as_mysql(self, compiler, connection, **extra_context):
-        # Use CONCAT_WS with an empty separator so that NULLs are ignored.
-        return super().as_sql(
-            compiler,
-            connection,
-            function="CONCAT_WS",
-            template="%(function)s('', %(expressions)s)",
-            **extra_context,
-        )
-
-    def coalesce(self):
-        # null on either side results in null for expression, wrap with coalesce
-        c = self.copy()
-        c.set_source_expressions(
-            [
-                Coalesce(expression, Value(""))
-                for expression in c.get_source_expressions()
+        # Systematically coalesce source expression to empty strings because
+        # their nullability cannot be accurately inferred.
+        if not connection.features.interprets_empty_strings_as_nulls:
```

Comment:
Happy to add a distinct feature flag for that if necessary. Oracle [didn't like](https://djangoci.com/job/pull-requests-oracle/database=oracle19,label=oracle,python=python3.11/672/testReport/junit/db_functions.text.test_concat/ConcatTests/test_basic/) the systematic coalescing and then pipe concatenation of expressions so since it's unnecessary for Oracle we avoid doing it in the first place.

```
ORA-12704: niezgodność zbioru znaków Help: https://docs.oracle.com/error-help/db/ora-12704/
```

I assume @felixxm configured CI with the Polish locale 😅

Reply:
It seems that this time only Oracle works out of the box :smile:

What do you think about using `as_sqlite()` and `as_postgresql()` instead of `as_sql()`?
```python
    def as_sqlite(self, compiler, connection, **extra_context):
        clone = self.copy()
        source_expressions = clone.get_source_expressions()
        clone.set_source_expressions([
            Coalesce(expr, Value(""))
            for expr in source_expressions
        ])  
        return clone.as_sql(compiler, connection, **extra_context)
        
    def as_postgresql(self, compiler, connection, **extra_context):
        clone = self.copy() 
        source_expressions = clone.get_source_expressions()
        clone.set_source_expressions([
            Coalesce(expr, Value(""))
            if isinstance(expr.output_field, (CharField, TextField))
            else Coalesce(Cast(expr, TextField()), Value(""))
            for expr in source_expressions
        ])  
        return clone.as_sql(compiler, connection, **extra_context)
```
We could avoid adding the new feature flag, and avoid calling `COALESCE()` for other databases (the current behavior).

Reply:
@felixxm not against it. The only downside I see is that it forces pretty much any third party backend to implement their own `as_{backend}` implementation as I believe the ANSII `||` implementation also requires coalesce to work as intended by `Concat`.

Reply:
As a concrete example, SQL server uses the `+` instead but still treats nulls like SQLite and Postgres but requires cast for Postgres

https://dbfiddle.uk/4divocCO
https://dbfiddle.uk/ojO2qG_M

If we use the `as_sql` approach all they would need to do to implement their `as_sqlserver` counter would be to set `requires_casted_text_in_pipes_concat = True` (maybe we need a better name that isn't tied to pipes, concat_op?) and then call `as_sql` with `arg_joiner=" + "`.

Reply:
Maybe instead of using feature flags we could add an `ops` hook :thinking: 
- default
```python
def prepare_concat_expr(self, expr):
    return Coalesce(expr, Value(""))
```
- Oracle
```python
def prepare_concat_expr(self, expr):
    return expr
```
- PostgreSQL
```python
def prepare_concat_expr(self, expr):
    if not isinstance(expr.output_field, (CharField, TextField)):
        expr = Cast(expr, TextField())
    return super().prepare_concat_expr(expr)
```
I would like to have something more "explicit" because a mix of feature flags for one function sounds like overkill. Moreover `interprets_empty_strings_as_nulls` is not related to `ORA-12704`, it's about different character sets, not about NULLs.

Reply:
An ops hook might be more appropriate yes. At the end of the day `Concat` is really just a combinator so I wonder if we could reuse `connection.ops.combine_expression` with a `CONCAT = '||'` without necessarily adding the `//` magic that @shangxiao alluded to on Trac. In other words, `Concat` would still need to be used but from the backend perspective all they would need is to adapt their `combine_expression` method to accept a new combinator?

> Moreover interprets_empty_strings_as_nulls is not related to ORA-12704, ORA-12704 is about different character sets, not about NULLs.

That I agree with 💯

Reply:
> In other words, `Concat` would still need to be used but from the backend perspective all they would need is to adapt their `combine_expression` method to accept a new combinator?

Sounds good!

Reply:
Looking a bit more into it I think this approach won't work as the signature of `combine_expression` is `(connection: str, sub_expressions: list[str])` so it cannot do any `Expression` introspection. That means that we can't avoid a systematic `::text` cast on Postgres.

I don't think it's a huge deal if we systematically cast to `::text` tbh but it seems that it was an irritant and caused us to add a test against it in 6364b6ee1071381eb3a23ba6b821fc0d6f0fce75.

I'm still going to push a commit that demonstrates how it would look like.

Reply:
Another discovery. `||` is already used by `contrib.postgres.search.SearchVectorCombinable` so we can't systematically cast to `::text` even if we wanted to.

If we want to use this approach we'll have to also change the nature of `combine_expression` to accept expressions I fear.

Reply:
> If we want to use this approach we'll have to also change the nature of combine_expression to accept expressions I fear.

That's unfortunate :disappointed: Are we going to use ops-hooks approach? or do you prefer using feature flags in `as_sql()`?